### DISCUSSION TITLE
IntelliJ setup instructions

### DISCUSSION THREAD

**Discussion by ajmacdonald** [voyanttools/Voyant#109] [General]
# IntelliJ Workspace Setup

### Prerequisites

- Git is installed
- Java 11

### Downloads

- Download IntelliJ IDEA Ultimate from [here](https://www.jetbrains.com/idea/download ). Note that only the Ultimate Edition includes the `Tomcat and TomEE` plugin. (I couldn't configure the free "Smart Tomcat" plugin for the Community Edition properly)

- Download [Apache Tomcat 9](https://archive.apache.org/dist/tomcat/tomcat-9/v9.0.96/bin/) (specifically the latest version `apache-tomcat-9.0.96.zip`) and extract its contents to a directory of your choice. We'll refer to this directory as `[TOMCAT]` in the instructions below.
  - To avoid potential permission issues later, ensure the proper permissions are set for the files. On macOS/Linux, run the command:
    ```
    chmod -R 755 [TOMCAT]/*
    ```

### IntelliJ Workspace Setup

1. In IntelliJ, go to `File > New > Project from Version Control` (ensure Git is selected). Enter the URL `https://github.com/voyanttools/trombone.git` and select a directory where the project should be stored.

2. Again, navigate to `File > New > Project from Version Control` (with Git selected). Enter the URL `https://github.com/voyanttools/Voyant.git` and choose the directory where this project should be stored. When prompted to choose where to open the `Voyant` project, select New Window.

3. Switch back to the `Trombone` project window in IntelliJ (as `Trombone` will be the main project). You may close the `Voyant` project window if you want to.

4. Go to `File > New > Module from Existing Sources...` and select the directory where you cloned the `Voyant` project.

5. In the Import Module dialog, select `Import module from external model > Maven`. After importing, both projects should appear in the Project explorer.

6. Open `IntelliJ > Settings ...` and navigate to `Build, Execution, Deployment > Compiler > Java Compiler`. Ensure that the `Target bytecode version` is set to `11` for both the `Voyant` and `trombone` modules. **Note:** You need to repeat this step whenever you change or sync Maven dependencies (i.e. `pom.xml`)

7. Go to `File > Project Structure...` and ensure that under `Project Settings > Project`, the `SDK` is set to Java 11.

8. Navigate to `Run > Edit Configurations...`, click the `+` button, and select `Tomcat Server > local`.

9. In the `Server` tab, under the `Application server` setting, click `Configure...` to manage your Tomcat installations.

10. Add a new installation by clicking `+` and selecting the `[TOMCAT]` directory as both the `Tomcat Home` and `Tomcat base` directory. If IntelliJ doesn't recognize the installation properly, ensure that permissions are set correctly and restart IntelliJ to apply the changes. Click `OK`.

11. In the `Application server` dropdown, select the Tomcat instance you just added.

12. Update the `URL` setting to `http://localhost:8080/voyant`.

13. Switch to the `Deployment` tab and verify that `Voyant:war exploded` is listed. If it's not, click the `+` icon, select `Artifact...`, and then choose `Voyant:war exploded`. Make sure to change the `Application context` path to `/voyant`. Click `OK` to save and close the dialog.

14. Finally, start the server by clicking the green "Run" button at the top (a green triangle) to start the configuration.

### Troubleshooting

Q: I see the error `java: package sun.security.action does not exist`
A: Ensure that the `SDK` is set correctly in the `Project Structure` dialog. Also, verify that both the `SDK` and `Target bytecode version` are set to Java 11.

**Comment by janisdd** [comment]
Here is the updated version. It's just inconvenient that we can't see the raw markdown from the comments.

# IntelliJ Workspace Setup

### Prerequisites

- Git is installed
- Java 11

### Downloads

- Download IntelliJ IDEA Ultimate from [here](https://www.jetbrains.com/idea/download ). Note that only the Ultimate Edition includes the `Tomcat and TomEE` plugin. (I couldn't configure the free "Smart Tomcat" plugin for the Community Edition properly)


- Download [Apache Tomcat 9](https://archive.apache.org/dist/tomcat/tomcat-9/v9.0.96/bin/) (specifically the latest version `apache-tomcat-9.0.96.zip`) and extract its contents to a directory of your choice. We'll refer to this directory as `[TOMCAT]` in the instructions below.
  - To avoid potential permission issues later, ensure the proper permissions are set for the files. On macOS/Linux, run the command:
    ```
    chmod -R 755 [TOMCAT]
    # or use quotes if the absolute path contains spaces
    chmod -R 755 "[TOMCAT]"
    ```


### IntelliJ Workspace Setup

1. In IntelliJ, go to `File > New > Project from Version Control` (ensure Git is selected). Enter the URL `https://github.com/voyanttools/trombone.git` and select a directory where the project should be stored.

![step1](https://github.com/user-attachments/assets/79353a38-6efb-4209-9841-b7319158ad38)

2. Again, navigate to `File > New > Project from Version Control` (with Git selected). Enter the URL `https://github.com/voyanttools/Voyant.git` and choose the directory where this project should be stored. When prompted to choose where to open the `Voyant` project, select New Window.

![step2](https://github.com/user-attachments/assets/cc1a8fe2-88d9-4665-9631-94a149b75cb1)

3. Switch back to the `Trombone` project window in IntelliJ (as `Trombone` will be the main project). You may close the `Voyant` project window if you want to.

4. Go to `File > New > Module from Existing Sources...` and select the directory where you cloned the `Voyant` project.

5. In the Import Module dialog, select `Import module from external model > Maven`. After importing, both projects should appear in the Project explorer.

![step3](https://github.com/user-attachments/assets/8e28bbfe-d010-46f4-9394-7d641dbe57ea)

6. Open `IntelliJ > Settings ...` and navigate to `Build, Execution, Deployment > Compiler > Java Compiler`. Ensure that the `Target bytecode version` is set to `11` for both the `Voyant` and `trombone` modules. **Note:** You need to repeat this step whenever you change or sync Maven dependencies (i.e. `pom.xml`)

![step4](https://github.com/user-attachments/assets/d8de2e63-083b-42ef-8616-d54020e24e8e)

7. Go to `File > Project Structure...` and ensure that under `Project Settings > Project`, the `SDK` is set to Java 11.

![step5](https://github.com/user-attachments/assets/f2a148d5-eed8-4c55-b0c5-e4fa7c5532d2)

8. Navigate to `Run > Edit Configurations...`, click the `+` button, and select `Tomcat Server > local`.

![step7](https://github.com/user-attachments/assets/885479d0-6fa6-4ee8-9c77-38dd0ad3d900)

9.  In the `Server` tab, under the `Application server` setting, click `Configure...` to manage your Tomcat installations.

![step8](https://github.com/user-attachments/assets/dc3d3dd0-7a27-4820-b5d0-21dbe50d0f61)

10.  Add a new installation by clicking `+` and selecting the `[TOMCAT]` directory as both the `Tomcat Home` and `Tomcat base` directory. If IntelliJ doesn't recognize the installation properly, ensure that permissions are set correctly and restart IntelliJ to apply the changes. Click `OK`.

![step9](https://github.com/user-attachments/assets/1160e521-ffc6-4b46-800e-9f34b5dfcf1c)
![step10](https://github.com/user-attachments/assets/cfc0ecdf-4906-4d61-8523-fef6b7a51f90)

11.  In the `Application server` dropdown, select the Tomcat instance you just added.

![step12](https://github.com/user-attachments/assets/f7cd62fb-93d8-4f49-a7e9-76c31ef1fe39)

12.  Update the `URL` setting to `http://localhost:8080/voyant`.

![step13](https://github.com/user-attachments/assets/369bfe60-3ccf-42bb-8d46-98519d510099)

13. Switch to the `Deployment` tab and verify that `Voyant:war exploded` is listed. If it's not, click the `+` icon, select `Artifact...`, and then choose `Voyant:war exploded`. Make sure to change the `Application context` path to `/voyant`. Click `OK` to save and close the dialog.

![step14](https://github.com/user-attachments/assets/731d286d-d882-4080-91e5-dcddacae6fd5)
![step15](https://github.com/user-attachments/assets/76913f5a-2e96-482e-8591-e7d08b6fca6c)
![step16](https://github.com/user-attachments/assets/30c55494-4ba9-4af2-94a5-b8a10e57bc4e)

14.  Finally, start the server by clicking the green "Run" button at the top (a green triangle) to start the configuration.

![step17](https://github.com/user-attachments/assets/c07bd76e-7588-4171-ba19-939a16993f0c)
![step18](https://github.com/user-attachments/assets/2b86a7d0-2473-4f5a-8037-57190bc56cf1)


### Troubleshooting

Q: I see the error `java: package sun.security.action does not exist`
A: Ensure that the `SDK` is set correctly in the `Project Structure` dialog. Also, verify that both the `SDK` and `Target bytecode version` are set to Java 11.

**Comment by ajmacdonald** [comment]
Hi @janisdd, thanks again for this. I've added this to the [wiki ](https://github.com/voyanttools/Voyant/wiki/IntelliJ-Workspace-Setup) and credited you.
